### PR TITLE
Ci self hosted sdk

### PR DESCRIPTION
**Pull request by DaMandal0rian**
This PR introduces the use of self hosted runners for the repository.

- supports Ubuntu-20.04, macOS-12 and Windows-server-2022 runners
- use cargo stable installed on servers
- removes windows protoc installation for self-hosted runner
- make runners compatible with self-hosted runners and github actions runners
- uses ubuntu-20.04 for backward compatibility with older glibc

### CODE REVIEW

File: .github/workflows/push.yml [hunk 1]
```diff
@@ -129,29 +139,30 @@ jobs:
...
   docs:
-    runs-on: ubuntu-22.04
+    runs-on: ${{ fromJson(github.repository_owner == 'subspace' && '["self-hosted", "ubuntu-20.04-x86-64", "Linux"]' || '["ubuntu-20.04"]') }}
     steps:
-      - uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8 # v3.1.0
+      - name: git checkout
+        uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8 # v3.1.0
 
       - name: Install Protoc
-        uses: arduino/setup-protoc@64c0c85d18e984422218383b81c52f8b077404d3 # v1.1.2
```
Comment (DaMandal0rian) [inline .github/workflows/push.yml:154]: version 1.1.2 fails for macOS with http 404 error.

File: .github/workflows/push.yml [hunk 2]
```diff
@@ -27,33 +27,37 @@ env:
...
+      # On macOS, we need a proper Clang version, not Apple's custom version without wasm32 support
       - name: Install LLVM and Clang
         uses: KyleMayes/install-llvm-action@8852e4d5c58653ed05135c0a5d949d9c2febcb00 # v1.6.1
         with:
           version: "15.0"
 
-      # Workaround to resolve link error with C:\msys64\mingw64\bin\libclang.dll
-      - name: Remove msys64
-        run: Remove-Item -LiteralPath "C:\msys64\" -Force -Recurse
-        if: runner.os == 'Windows'
```
Comment (nazar-pc) [inline .github/workflows/push.yml:56]: Why was this removed?

File: .github/workflows/push.yml [hunk 3]
```diff
@@ -146,12 +151,13 @@ jobs:
         run: cargo sweep --maxsize ${{ env.MAX_TARGET_SIZE }}
 
   docs:
-    runs-on: ubuntu-22.04
+    runs-on: ubuntu-latest
```
Comment (nazar-pc) [inline .github/workflows/push.yml:154]: Why change to latest?

File: .github/workflows/rustdoc.yml [hunk 4]
```diff
@@ -15,24 +15,26 @@ env:
...
 
+      - name: Install Protoc
+        uses: arduino/setup-protoc@149f6c87b92550901b26acd1632e11c3662e381f # v1.3.0
+        with:
+          repo-token: ${{ secrets.GITHUB_TOKEN }}
+
       # Build the rust crate docs
       # Use `RUSTC_BOOTSTRAP` in order to use the `--enable-index-page` flag of rustdoc
       # This is needed in order to generate a landing page `index.html` for workspaces
+
```
Comment (nazar-pc) [inline .github/workflows/rustdoc.yml:37]:
```suggestion
```

**Review by nazar-pc — COMMENTED** [review]
Also needs to rebase and squash everything

File: .github/workflows/push.yml (hunk 3 above)
Comment (DaMandal0rian) [reply .github/workflows/push.yml:154]: 22.04 is the [latest](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources)

File: .github/workflows/push.yml (hunk 2 above)
Comment (DaMandal0rian) [reply .github/workflows/push.yml:56]: @nazar-pc the job only runs on linux so don't need it.

File: .github/workflows/push.yml (hunk 3 above)
Comment (nazar-pc) [reply .github/workflows/push.yml:154]: It is today, in a bit over half a year it will not be. Please use specific versions to avoid surprises in the future.

File: .github/workflows/push.yml (hunk 2 above)
Comment (nazar-pc) [reply .github/workflows/push.yml:56]: Maybe it should be running on all platforms though? I suspect it was too slow before, but we do want to know that SDK works on all 3 platforms.

**Review by nazar-pc — APPROVED** [review]

**Review by nazar-pc — DISMISSED** [review]
Let's squash-merge this, too many commits for such a small change
